### PR TITLE
Initial Cajita tutorial examples

### PR DESCRIPTION
**Pull request by streeve**
Part of #352

### CODE REVIEW

**Review by sslattery — APPROVED** [review]
Just a few small things. Thanks for getting a start on a Cajita tutorial!

File: example/cajita_tutorial/01_types/types_example.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+//---------------------------------------------------------------------------//
+// Types example.
+//---------------------------------------------------------------------------//
+void typesExample()
+{
+    /*
+      Cajita types are simple structs to used build grid and mesh objects. This
+      includes dimension tags (i,j,k) and mesh entity tags: cell, node, edge,
+      and face. In addition there are simple types to differeniate types of
```
[the remark this answers is not in the document]
Comment (streeve) [reply example/cajita_tutorial/01_types/types_example.cpp]: Fixed

File: example/cajita_tutorial/01_types/types_example.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    */
+    std::cout << "Is Cell a Cell? " << Cajita::isCell<Cajita::Cell>()
+              << std::endl;
+    std::cout << "Is Node a Node? " << Cajita::isNode<Cajita::Node>()
+              << std::endl;
+    std::cout << "Is Cell a Node? " << Cajita::isNode<Cajita::Cell>() << "\n"
+              << std::endl;
+
+    /*
+      Each mesh cell also has twelve "Edges" and six "Faces" which are indexed
```
[the remark this answers is not in the document]
Comment (streeve) [reply example/cajita_tutorial/01_types/types_example.cpp]: Added 2D details

File: example/cajita_tutorial/01_types/types_example.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    std::cout << "Is I Edge an Edge? "
+              << Cajita::isEdge<Cajita::Edge<Cajita::Dim::I>>() << std::endl;
+    std::cout << "Is J Face a Face? "
+              << Cajita::isFace<Cajita::Face<Cajita::Dim::J>>() << std::endl;
+    std::cout << "Is K Face an Edge? "
+              << Cajita::isEdge<Cajita::Face<Cajita::Dim::K>>() << "\n"
+              << std::endl;
+
+    /*
+      There are also type tags to distinguish local and global indexing: local
```
[the remark this answers is not in the document]
Comment (streeve) [reply example/cajita_tutorial/01_types/types_example.cpp]: Replaced with your description

File: example/cajita_tutorial/02_global_mesh/global_mesh_example.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+#include <iostream>
+
+//---------------------------------------------------------------------------//
+// Types example.
+//---------------------------------------------------------------------------//
+void globalMeshExample()
+{
+    /*
+      All meshes in Cajita are logically recitlinear: each mesh cell is either
```
[the remark this answers is not in the document]
Comment (streeve) [reply example/cajita_tutorial/02_global_mesh/global_mesh_example.cpp:22]: Added 2D

File: example/cajita_tutorial/02_global_mesh/global_mesh_example.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    /*
+      Now create the global mesh with the low and high corners and total cells.
+    */
+    auto global_mesh = Cajita::createUniformGlobalMesh(
+        global_low_corner, global_high_corner, global_num_cell );
+
+    /*
+      Instead create the global mesh with the cell size.
+    */
+    global_mesh = Cajita::createUniformGlobalMesh(
```
[the remark this answers is not in the document]
Comment (streeve) [reply example/cajita_tutorial/02_global_mesh/global_mesh_example.cpp]: Done

File: example/cajita_tutorial/03_partitioner/partitioner_example.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        std::cout << "Cajita Partitioner Example" << std::endl;
+        std::cout << "    (intended to be run with MPI)\n" << std::endl;
+    }
+
+    /*
+      The simpler Cajita partitioner, DimBlockPartitioner finds the
+      decomposition closest to uniform internally using MPI. It is templated on
+      spatial dimension.
+    */
+    Cajita::DimBlockPartitioner<3> auto_partitioner;
```
[the remark this answers is not in the document]
Comment (streeve) [reply example/cajita_tutorial/03_partitioner/partitioner_example.cpp]: Done

File: example/cajita_tutorial/03_partitioner/partitioner_example.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    /*
+      Extract the MPI ranks per spatial dimension. The second argument, global
+      cells per dimension, is unused in this case.
```
[the remark this answers is not in the document]
Comment (streeve) [reply example/cajita_tutorial/03_partitioner/partitioner_example.cpp]: Added

File: example/cajita_tutorial/03_partitioner/partitioner_example.cpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        for ( int d = 0; d < 3; ++d )
+            std::cout << ranks_per_dim_3[d] << " ";
+        std::cout << std::endl;
+    }
+
+    /*
+      Instead, the manual variant, ManualBlockPartitioner, accepts a user
+      decomposition (the number of MPI ranks per dimension).
+
+      Here, MPI is still used to set the initial ranks per dimension to ensure
```
[the remark this answers is not in the document]
Comment (streeve) [reply example/cajita_tutorial/03_partitioner/partitioner_example.cpp]: Good point - this is much clearer

File: example/cajita_tutorial/03_partitioner/partitioner_example.cpp [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    MPI_Comm_rank( MPI_COMM_WORLD, &comm_rank );
+
+    if ( comm_rank == 0 )
+    {
+        std::cout << "Cajita Partitioner Example" << std::endl;
+        std::cout << "    (intended to be run with MPI)\n" << std::endl;
+    }
+
+    /*
+      The simpler Cajita partitioner, DimBlockPartitioner finds the
```
[the remark this answers is not in the document]
Comment (streeve) [reply example/cajita_tutorial/03_partitioner/partitioner_example.cpp]: Added

**Comment by streeve** [comment]
@sslattery ready to merge?

**Pull request merged**